Getting rid of VerilogA

ojarvin opened the pull request:
Spectre runs VerilogA-free in this branch now.

Tested for spice_iofile types
- event (was already done)
- sample, for both digital (ioformat='bin' or 'dec') and voltage (ioformat='volt')
- time

Each sample (each bit in a bus) or time signal gets saved as an event-type print from spectre, and all further parsing is done in Python. This should make it quite trivial to extend to Eldo and ngspice as well (as long as they support event IOs without veriloga). Downside is again larger output filesize.

I also deprecated spice_iofile type 'vsample' because in my opinion it is a part of the 'sample' class of IOs which are defined by having discrete time points. The 'vsample' functionality is usable with 'sample' IO and ioformat='volt'. This also makes it quite trivial to extend this to Eldo and ngspice, which (AFAIK) didn't previously support sampled analog signals.

Once the features are implemented for Eldo and ngspice as well, this pull-request closes #4.

mkosunen commented:
Is it this IOtype Volt that needs to be implemented for eldo and NGspice?

ojarvin commented:
Yes, I can do it. To clarify, the IO types are

* event
    - Implemented for spectre, eldo and ngspice
* time
    - Implemented for spectre. Eldo and ngspice I'm not sure, but it can be done.
* sample
    - ioformat = 'bin', implemented for spectre. Eldo and ngspice I would need to check.
    - ioformat = 'dec', same as above.
    - ioformat = 'volt', implemented for spectre. This replaces the old iotype 'vsample'. This should maybe be renamed to something that hints at "discrete time, continuous amplitude". I think 'sample' as iotype already tells that it's at least discrete time, but unlike 'bin' and 'dec', this 'volt' thing doesn't get rounded to bits.
    
If it seems too unclear like this, I can separate the analog sampled type into its own iotype, they just share quite a bit of code right now.

mkosunen commented:
About ioformat: think sample means discrete in time, but no quantization in amplitude. I

What is the functionality we try to achieve with bin, dec, and volt, is volt actually float? (in rtl, formats  are int, sint, complex and  scomplex, I think) 

Anyway, let's keep it like this for now, and open a new issue and PR if needed. I will do some simple tests and merge.

mkosunen commented:
I do not know if this is related to changes, but ngspice sim fails because of fatal error determined by Python execution 
`ERROR: (internal)  This operation is not defined for display type PrinterOnly.
Can't open viewport for graphics.
ERROR: (internal)  This operation is not defined for display type PrinterOnly.
Can't open viewport for graphics.
ERROR: (internal)  This operation is not defined for display type PrinterOnly.
Can't open viewport for graphics.
ERROR: (internal)  This operation is not defined for display type PrinterOnly.
Can't open viewport for graphics.
Note: No ".plot", ".print", or ".fourier" lines; no simulations run
`
The case was the same also earlier, but now python considers it fatal.

Cause for this is that there is a plot commnad in the testbench and the simulation is run in batch mode. One possible solution would be to make those plots conditional to 'interactive_spice' parameter.

@ojarvin Do you observe this behaviour?

sporrasm commented:
@mak @ojarvin there was a problem reading some event-type outputs. Spectre seems to use ":" character as a delimiter for instance pins. This causes the reading to file since we split the matched line by ":" character. I had to change the grep default delimiter to some other character ("|"). I pushed the fix to this branch since the ngspice event types are handled in the same function. Do you know if ngspice uses "|" delimiter in some case in the print output? If not, commit 68d0e4e should fix the problem.

ojarvin commented:
> I do not know if this is related to changes, but ngspice sim fails because of fatal error determined by Python execution `ERROR: (internal) This operation is not defined for display type PrinterOnly. Can't open viewport for graphics. ERROR: (internal) This operation is not defined for display type PrinterOnly. Can't open viewport for graphics. ERROR: (internal) This operation is not defined for display type PrinterOnly. Can't open viewport for graphics. ERROR: (internal) This operation is not defined for display type PrinterOnly. Can't open viewport for graphics. Note: No ".plot", ".print", or ".fourier" lines; no simulations run ` The case was the same also earlier, but now python considers it fatal.
> 
> Cause for this is that there is a plot commnad in the testbench and the simulation is run in batch mode. One possible solution would be to make those plots conditional to 'interactive_spice' parameter.
> 
> @ojarvin Do you observe this behaviour?

Ah I added a simulator status code check which throws a fatal error on non-zero code. This makes sense for Spectre and Eldo, but apparently ngspice throws errors even in the normal case. Should I make an exception for ngspice in the code or remove this check?

mkosunen commented:
Did it already, just haven't pushed it. I tru to do it once I get away from the meeitng.

sporrasm commented on spice/testbench.py:
```diff
@@ -287,27 +283,26 @@ def dcsourcestr(self):
         if not hasattr(self,'_dcsourcestr'):
             self._dcsourcestr = "%s DC sources\n" % self.parent.syntaxdict["commentchar"]
             for name, val in self.dcsources.Members.items():
+                supply = '%s%s' % (val.sourcetype.upper(),val.name.upper())
```
@ojarvin Is this strictly necessary to make things work with ngspice/eldo? Previously the supply name was V<val.name> or I<val.name>, with name in lower case. This change of course breaks all testbench entities, which follow the old naming
convention. If this is strictly required for unification of all simulators, I will make the necessary changes to my testbenches.

ojarvin replied:
Well, not strictly necessary. However, I think this new method is less confusing and more consistent with everything. The problem is that the user gives a name "supply" but they should somehow now to look for "VSUPPLY" in the extract dictionary later. I guess this can be fixed by storing the extracted powers with 'name' directly, while the testbench uses the old naming...

sporrasm replied:
I see, this seems logical. It is better that the names are consistent across the board. No worries, I will change my testbenches to the new convention.

mkosunen replied:
My two cents: case conversions should be avoided whenever possible, and the assumtiont should be that things are case sensitive.

ojarvin replied:
Agreed. However, our friend Eldo automatically uppercases everything in the print files, which makes it impossible to search for them with user-given cases. Right now all cases should be preserved, except when they are used as keys internally (at least that was the intention).

ojarvin commented:
@mkosunen @sporrasm I added the sphinx-generated PDF file to the repo and linked it to the README.md. Do you think this is accessible-enough way to provide documentation? If you open https://github.com/TheSystemDevelopmentKit/spice/tree/no_veriloga and click the big blue "Documentation", it does what one might expect.

👍 or 👎?

mkosunen commented:
@ojarvin , please remove the status check. I removed the plots from the batch mode, but the problem is actually this:
Note: No ".plot", ".print", or ".fourier" lines; no simulations run

We should have a plot that plots nothing. This must be a bug, and I think I will report this to ngspice developers.

ojarvin commented:
> @ojarvin , please remove the status check. I removed the plots from the batch mode, but the problem is actually this: `Note: No ".plot", ".print", or ".fourier" lines; no simulations run`
> 
> We should have a plot that plots nothing. This must be a bug, and I think I will report this to ngspice developers.

Can you test now? I changed the check from 'fatal' to 'error' and it should not quit now.

sporrasm commented:
@mkosunen @ojarvin Everything seems to be working a-okay. Green light for merge from me.
